### PR TITLE
Introduce a filter to control whether an order is partially refunded in `wc_create_refund`

### LINKED ISSUE #31416: Introduce a filter to control whether an order is partially refunded in `wc_create_refund`
### Describe the solution you'd like

Hey team!

While trying to support admin refunds in the WooCommerce Gift Cards extension, we needed to reuse the `wc_create_refund()` function. This function includes a hard check on whether the order is considered partial or fully refunded. See https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/wc-order-functions.php#L613

This check creates an issue for us when an order is considered fully refunded, but there is still some captured gift card left to refund.

It would be great to introduce a filter to control whether an order is partially refunded or not in `wc_create_refund`.

Related issue: https://github.com/woocommerce/woocommerce/issues/31413

### Describe alternatives you've considered

Please visit the P2 project thread for more details: https://wp.me/pdqkbG-6O#comment-169

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by xristos3490**
Closes #31416 

### Specification 

Please check the issue description for additional details.

### CODE REVIEW

**Comment by manospsyx** [comment]
@woocommerce/proton see pdqkbG-6O-p2#comment-173

**Comment by splaquet** [comment]
How about considering another Order Status of "Partial Refund"?

We've implemented a few custom order statuses, with one being Partial Refund.  We were finding that Financial Reports, ShipStation, and other aspects were becoming confusing for much of our staff.  The Partial Refund signifies that money still came in, and also allows us to easily (re)view past Partial Refunds.  

While we've found several advantages to this additional order status, we (as techs) can feel more confident that the big wheel can keep rolling, and Accounting Reports & Shipments are still being produced correctly.

**Comment by jeffstieler** [comment]
I've got a question about the implementation - posted to the source issue. I think we should probably discuss this as well @vedanshujain.

File: plugins/woocommerce/includes/wc-order-functions.php [hunk 1]
```diff
@@ -610,7 +610,7 @@ function wc_create_refund( $args = array() ) {
 			}
 
 			// Trigger notification emails.
-			if ( ( $remaining_refund_amount - $args['amount'] ) > 0 || ( $order->has_free_item() && ( $remaining_refund_items - $refund_item_count ) > 0 ) ) {
+			if ( (bool) apply_filters( 'woocommerce_order_is_partially_refunded', ( $remaining_refund_amount - $args['amount'] ) > 0 || ( $order->has_free_item() && ( $remaining_refund_items - $refund_item_count ) > 0 ), $order->get_id(), $refund->get_id() ) ) {
```
Comment (barryhughes) [inline plugins/woocommerce/includes/wc-order-functions.php:623]: Please let's document the filter and give a hint at when it should/should not be used/what the implications are, and capture the three `@param` types: then I think this is good to go.

**Review by barryhughes — CHANGES_REQUESTED** [review]

File: plugins/woocommerce/includes/wc-order-functions.php (hunk 1 above)
Comment (xristos3490) [reply plugins/woocommerce/includes/wc-order-functions.php:623]: Thanks for the review, @barryhughes! Just added the doc block.

**Comment by botwoo** [comment]
:bar_chart: Test reports for this pull request have been published and are accessible through the following links:

- [REST API Test Report](https://woocommerce.github.io/woocommerce-test-reports/pr/31417/api/)
- E2E Test Report (TBC)

Latest commit referenced in the reports: `Update @version to 6.7.0.` 277e3635aba5a271e21ba59550637a4b0b6488ab
This comment will automatically be updated with the latest referenced commit when you push new changes to this pull request.

---

_Visit the [WooCommerce Test Reports](https://woocommerce.github.io/woocommerce-test-reports) homepage to view all published reports. See the [FAQs](https://github.com/woocommerce/woocommerce-test-reports/wiki/FAQs) page if you're having problems accessing them._

**Review by barryhughes — APPROVED** [review]

**Pull request merged**